include/kernel.h

```diff
@@ -1431,7 +1431,7 @@ struct k_timer {
 	/* user-specific data, also used to support legacy features */
 	void *user_data;
 
-	_OBJECT_TRACING_NEXT_PTR(k_timer);
+	_OBJECT_TRACING_NEXT_PTR(k_timer)
```

Comment:
It would probably be more clearer, if the commit message would explain why this change is needed. Now it just states what is being done but that is quite obvious by looking the code. The real issue is that if the macro is not defined, then there is extra semicolon inside the struct.

Reply:
isnt that in the commit message already?

`ISO C99 does not allow extra ‘;’ outside of a function`

Reply:
Couldn't this confusing (and unusual) inclusion of a semicolon in the macro definition be avoided by having it expand to a dummy zero-length array element instead of nothing?

Reply:
> isnt that in the commit message already?

> ISO C99 does not allow extra ‘;’ outside of a function

That commit message is possibly wrong (doesn't tell the whole truth).

Reply:
> Couldn't this confusing (and unusual) inclusion of a semicolon in the macro definition

What's so confusing and unusual? In normal languages, macros are Turing-complete, imagine how that is "confusing". And in C, I didn't hear someone telling that the preprocessor as defined by some standard disallows "inclusion" of semicolon in macro body, references are welcome.

Reply:
Ok, @jhedberg explains the issue here: https://github.com/zephyrproject-rtos/zephyr/pull/9917#discussion_r217603899 (too bad we have 2 branches of discussion of the same issue).

Which only shows that there can't be too detailed commit message, and majority of commit messages are terse up to confusion and doesn't described the changes they do well.

Reply:
> I didn't hear someone telling that the preprocessor as defined by some standard disallows "inclusion" of semicolon in macro body, references are welcome.

@pfalcon I didn't say that it's disallowed. It's just that most macros in the code base don't do this, which to the casual reader makes it look like there's a bug in the code that's using the macro (since one expects to see a semicolon at the end of each struct member declaration). Anyway, I'm curious to hear thoughts on my zero-length array proposal that should help avoid all this.

Reply:
> with which compiler issues fixed in this PR were discovered?

Ok, the description says "permits code to compile with our Clang-based compiler.". Is that public compiler? Based on recent Clang?

Anyway, I guess it's up to @mped-oticon to look at zero-sized array solution and test if it works/see if he'd be willing to use it.

(My concern with this PR overall remains not detailed enough commit messages.)

Reply:
@jhedberg zero-length array (aka flexible array member) is a fine idea/hack but has caveats: Look at `kernel.h`'s `struct k_stack` (or k_msgq or k_pipe), is the last member is some `u8_t flags`. That means our zero-length array is not the last member. This violates C99, see [1].

I would also be very skeptical of nesting structs that contain flexible array members - as the address for subsequent members would not be known at compile-time (which is required). I don't know if these struct nest though, but even if not, it could blow up later.

[1]
C99 spec: "As a special case, the **_last_** element of a structure with more than one named member may have an incomplete array type; this is called a flexible array member."

Reply:
@pfalcon 
> [...] looks confusing IMHO. So, doing reasonable steps to preserve them makes sense IMHO.
Hopefully the ALL_CAPS and parenthesis should inform you that this line is a macro, and therefore caution should be exercised. I agree it would be nice to keep up appearence, but IMHO the benefit (portability) outweighs the cons (readabillity). A popular RTOS targetting embedded systems will have ugly corners. Good thing we're not in a beauty contest :)

Reply:
@pfalcon I will add make the commit messages more detailed.

Reply:
@mped-oticon I wasn't aware that this wasn't formally allowed elsewhere than at the end of structs. The main reason I thought it was is that there's a major project that uses it, namely the Linux kernel:

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/skbuff.h#n737
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/skbuff.h#n750
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/skbuff.h#n761

There are probably many more occurrences, but I pointed out these from the subsystem that I'm most familiar with. Anyway, because Linux uses zero-length arrays like this, it would seem bad business for a compiler not to support it, so I wonder how likely it is that we'd run into issues.

Reply:
>  I will make the commit messages more detailed

Thanks, appreciated!

Reply:
@pfalcon, @jhedberg It may work out with [0], but [] definitely does not. I thought zero-sized array and flexible arrays were treated the same, but I guess not :)

https://gcc.godbolt.org/z/1WEeEq 
removing `-pedantic` quenches the warning. So I guess [0] is fine if we never want `-pedantic` with `-Wall`, which would be a PITA.

That leaves me satisfied with the [0] proposal.